activerecord/test/cases/adapters/mysql2/virtual_column_test.rb

```diff
@@ -0,0 +1,59 @@
+require "cases/helper"
+require "support/schema_dumping_helper"
+
+if ActiveRecord::Base.connection.supports_virtual_columns?
+  class Mysql2VirtualColumnTest < ActiveRecord::Mysql2TestCase
+    include SchemaDumpingHelper
+
+    self.use_transactional_tests = false
+
+    class VirtualColumn < ActiveRecord::Base
+    end
+
+    def setup
+      @connection = ActiveRecord::Base.connection
+      @connection.create_table :virtual_columns, force: true do |t|
+        t.string  :name
+        t.virtual :upper_name,  type: :string,  as: "UPPER(name)"
+        t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
+      end
+      VirtualColumn.create(name: "Rails")
+    end
+
+    def teardown
+      @connection.drop_table :virtual_columns, if_exists: true
+      VirtualColumn.reset_column_information
+    end
+
+    def test_virtual_column
+      column = VirtualColumn.columns_hash["upper_name"]
+      assert column.virtual?
+      assert_match %r{\bVIRTUAL\b}, column.extra
+      assert_equal "RAILS", VirtualColumn.take.upper_name
+    end
+
+    def test_stored_column
+      column = VirtualColumn.columns_hash["name_length"]
+      assert column.virtual?
+      assert_match %r{\b(?:STORED|PERSISTENT)\b}, column.extra
+      assert_equal 5, VirtualColumn.take.name_length
+    end
+
+    def test_change_table
+      @connection.change_table :virtual_columns do |t|
+        t.virtual :lower_name, type: :string, as: "LOWER(name)"
+      end
+      VirtualColumn.reset_column_information
+      column = VirtualColumn.columns_hash["lower_name"]
+      assert column.virtual?
+      assert_match %r{\bVIRTUAL\b}, column.extra
+      assert_equal "rails", VirtualColumn.take.lower_name
+    end
+
+    def test_schema_dumping
```

Comment:
I confirmed MySQL 8.0.0 always quotes column names. So I'll change the definition in the tests includes back quotes. Thanks!

```
root@localhost [test] > CREATE TABLE `virtual_columns` (`id` bigint AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `upper_name` varchar(255) AS (UPPER(name)), `name_length` int AS (LENGTH(name)) STORED) ENGINE=InnoDB;
Query OK, 0 rows affected (0.01 sec)

root@localhost [test] > show create table virtual_columns\G                                                                                 *************************** 1. row ***************************
       Table: virtual_columns
Create Table: CREATE TABLE `virtual_columns` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `upper_name` varchar(255) GENERATED ALWAYS AS (upper(`name`)) VIRTUAL,
  `name_length` int(11) GENERATED ALWAYS AS (length(`name`)) STORED,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
1 row in set (0.00 sec)

root@localhost [test] > select column_name, column_type, generation_expression from information_schema.columns where table_schema = database() and table_name = 'virtual_columns';
+-------------+--------------+-----------------------+
| COLUMN_NAME | COLUMN_TYPE  | GENERATION_EXPRESSION |
+-------------+--------------+-----------------------+
| id          | bigint(20)   |                       |
| name        | varchar(255) |                       |
| name_length | int(11)      | length(`name`)        |
| upper_name  | varchar(255) | upper(`name`)         |
+-------------+--------------+-----------------------+
4 rows in set (0.00 sec)

root@localhost [test] > select @@version;
+---------------+
| @@version     |
+---------------+
| 8.0.0-dmr-log |
+---------------+
1 row in set (0.00 sec)
```